src/generic/listctrl.cpp

```diff
@@ -151,6 +151,34 @@ wxListItemData::wxListItemData(wxListMainWindow *owner)
         m_rect = new wxRect;
 }
 
+// Check if the item is visible
+bool wxGenericListCtrl::IsVisible(long item) const
+{
+    wxRect itemRect, rect;
+    GetItemRect( item, itemRect );
+    rect = GetClientRect();
+/*    if( itemRect.y + itemRect.height <= 0 || itemRect.GetTop() > rect.GetBottom() )
+        return false;
+    else
+    {
+        wxRect headerRect( 0, 0, rect.GetWidth(), 0 );
+        if( m_headerWin )
+            headerRect = m_headerWin->GetRect();
+        if( itemRect.Intersects( headerRect ) )
+            return false;
+        else
+            return true;
+    }*/
+    bool visible = rect.Intersects( itemRect );
+    if( visible && m_headerWin )
+    {
+        wxRect headerRect = m_headerWin->GetClientRect();
+        headerRect.height++;
```

Comment:
Why do we do this? It merits at least a comment.

Reply:
I suggested this so its fair I answer. We need to correct for the `+1` added in `GetSubItemRect`.

For example (without adding +1), the header rect is `0+24` (y+height) and the item rect behind the header is `5+20` which would mean it is visible, while in fact it is not. For consistency it might be better to use `rect.y--` but the effect is the same.

Reply:
Sorry, I'm not sure if I follow the logic: in this example, the last header pixel is `y=23`, while the bottom pixel of the item is at `y=24`, so it looks like it ought to be considered "visible" if our criterion for visibility is that at least one item pixel is visible.

Of course, thinking about it now, it's not clear at all if this is the good criterion. Maybe we should rather check that the item is _entirely_ visible?

Reply:
But the last pixel is not visible. The last header pixel is `24`, the last `GetItemRect` pixel is `25` (including the `+1`). The actual position of the last item pixel is `24`, making the item hidden behind the header.

The native Windows implementation uses partial visibility, so does the `IsVisible` implementation in `wxTreeCtrl`. I think we should not change it.

Reply:
There could be an off by 1 error somewhere in `RECT` to/from `wxRect` conversions... Because for `wxRect` with `y=0` and `height=24`, its last pixel is supposed to be `23`, i.e. `y + height` is exclusive (which is why `wxRect::GetBottom()` returns `y + height - 1`). But MSW `RECT` is inclusive, so it really looks like a couple of `+1`s are missing in `wxRectFromRECT()`.

How did we fail to notice this before (the error seems to have existed since ed7919869508ccd5c868683c28706f82e82a1b9b...) and where else did we compensate for this by adding or subtracting 1 is another question :-(

Reply:
This is the generic `wxListCtrl`, it doesn't use `RECT`.

That commit seems unrelated to me? I tracked to 7e6874c01461f87d3ac7e454973a91d743942e4c.

Reply:
> This is the generic `wxListCtrl`, it doesn't use `RECT`.

Oops, it's really time to go to sleep for me, sorry. I am still not sure everything is correct here neither, but perhaps I should get back to it tomorrow.

Reply:
Hi,
Is there anything left here in this piece that needs fixing?

Thx.

Reply:
No fixing (or change it to `y++`). Just add a comment why `height++` is needed (e.g.: `take into account the +1 added in GetItemRect()`).